Mathlib/NumberTheory/NumberField/FinitePlaces.lean

```diff
@@ -0,0 +1,240 @@
+    instFieldAdicCompletion, Valued.norm, Valuation.RankOne.hom, MonoidWithZeroHom.coe_mk,
+    ZeroHom.coe_mk, embedding, UniformSpace.Completion.coeRingHom, RingHom.coe_mk, MonoidHom.coe_mk,
+    OneHom.coe_mk, Valued.valuedCompletion_apply, toNNReal_Valued_eq_vadicAbv]
+
+/-- The norm of the image after the embedding associated to `v` is equal to the norm of `v` raised
+to the power of the `v`-adic valuation. -/
+theorem FinitePlace.norm_def' (x : K) : ‖embedding v x‖ = toNNReal (norm_ne_zero v)
+    (v.valuation x) := by
+  rw [norm_def, vadicAbv_def]
+
+/-- The norm of the image after the embedding associated to `v` is equal to the norm of `v` raised
+to the power of the `v`-adic valuation for integers. -/
+theorem FinitePlace.norm_def'' (x : 𝓞 K) : ‖embedding v x‖ = toNNReal (norm_ne_zero v)
+    (v.intValuationDef x) := by
+  rw [norm_def, vadicAbv_def, valuation_eq_intValuationDef]
+
+open FinitePlace
+
+/-- The `v`-adic norm of an integer is at most 1. -/
+theorem norm_le_one (x : 𝓞 K) : ‖embedding v x‖ ≤ 1 := by
+  rw [norm_def', NNReal.coe_le_one, toNNReal_le_one_iff (one_lt_norm v)]
+  exact valuation_le_one v x
+
+/-- The `v`-adic norm of an integer is 1 if and only if it is not in the ideal. -/
+theorem norm_eq_one_iff_not_mem (x : 𝓞 K) : ‖(embedding v) x‖ = 1 ↔ x ∉ v.asIdeal := by
+  rw [norm_def'', NNReal.coe_eq_one, toNNReal_eq_one_iff (v.intValuationDef x)
+    (norm_ne_zero v) (one_lt_norm v).ne', ← dvd_span_singleton,
+    ← intValuation_lt_one_iff_dvd, not_lt]
+  exact (intValuation_le_one v x).ge_iff_eq.symm
+
+/-- The `v`-adic norm of an integer is less than 1 if and only if it is in the ideal. -/
+theorem norm_lt_one_iff_mem (x : 𝓞 K) : ‖embedding v x‖ < 1 ↔ x ∈ v.asIdeal := by
+  rw [norm_def'', NNReal.coe_lt_one, toNNReal_lt_one_iff (one_lt_norm v),
+    intValuation_lt_one_iff_dvd]
+  exact dvd_span_singleton
+
+end FinitePlace
+
+namespace FinitePlace
+variable {K : Type*} [Field K] [NumberField K]
+
+instance : FunLike (FinitePlace K) K ℝ where
+  coe w x := w.1 x
+  coe_injective' _ _ h := Subtype.eq (AbsoluteValue.ext <| congr_fun h)
+
+instance : MonoidWithZeroHomClass (FinitePlace K) K ℝ where
+  map_mul w := w.1.map_mul
+  map_one w := w.1.map_one
+  map_zero w := w.1.map_zero
+
+instance : NonnegHomClass (FinitePlace K) K ℝ where
+  apply_nonneg w := w.1.nonneg
+
+@[simp]
+theorem apply (v : HeightOneSpectrum (𝓞 K)) (x : K) : mk v x =  ‖embedding v x‖ := rfl
+
+/-- For a finite place `w`, return a maximal ideal `v` such that `w = finite_place v` . -/
+noncomputable def maximalIdeal (w : FinitePlace K) : HeightOneSpectrum (𝓞 K) := w.2.choose
+
+@[simp]
+theorem mk_maximalIdeal (w : FinitePlace K) : mk (maximalIdeal w) = w := Subtype.ext w.2.choose_spec
+
+@[simp]
+theorem norm_embedding_eq (w : FinitePlace K) (x : K) :
+    ‖embedding (maximalIdeal w) x‖ = w x := by
+  rw [show w x = (mk (maximalIdeal w)) x by simp only [mk_maximalIdeal], apply]
```

Comment:
```suggestion
  conv_rhs => rw [← mk_maximalIdeal w, apply]
```
`conv` and friends allows you more targeted rewrites (and more).

Reply:
I didn't know of this tactic! It's quite useful!